From: Philipp Hachtmann <phacht@linux.vnet.ibm.com>
Subject: s390/watchdog: Convert and enhance the s390 watchdog
Patch-mainline: not yet
References: bsc#934542,FATE#318029,LTC#KRN1309

Summary:     s390/watchdog: Convert and enhance the s390 watchdog
Description: Converted the vmwatchdog driver to use the kernel's watchdog API.
             It is now also called diag288_wdt instead of vmwatchdog.
             The new driver now also supports operation under LPAR.

Upstream-Description:

             (not upstream) s390/watchdog: add kernel messages for KRN1309

Signed-off-by: Philipp Hachtmann <phacht@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 Documentation/kmsg/s390/diag288_wdt |  117 ++++++++++++++++++++++++++++++++++++
 Documentation/kmsg/s390/vmwatchdog  |   25 -------
 2 files changed, 117 insertions(+), 25 deletions(-)

--- /dev/null
+++ b/Documentation/kmsg/s390/diag288_wdt
@@ -0,0 +1,117 @@
+/*?
+ * Text: "The watchdog was deactivated\n"
+ * Severity: Informational
+ * Description:
+ * Diagnose instruction 0x288 was called to activate the diag288 watchdog.
+ * The diagnose call returned an error that cannot be handled by the device driver.
+ * The watchdog stays inactive.
+ * User action:
+ * Contact your support organization.
+ */
+
+/*?
+ * Text: "The watchdog cannot be activated\n"
+ * Severity: Error
+ * Description:
+ * Diagnose instruction 0x288 was called to activate the diag288 watchdog.
+ * The diagnose call returned an error that cannot be handled by the device driver.
+ * The watchdog stays inactive.
+ * User action:
+ * Contact your support organization.
+ */
+
+/*?
+ * Text: "The watchdog cannot be initialized\n"
+ * Severity: Error
+ * Description:
+ * Diagnose instruction 0x288 was called to initialize the diag288 watchdog.
+ * The diagnose call returned an error that cannot be handled by the device driver.
+ * The watchdog stays inactive.
+ * A possible reason for this error is that your real or virtual hardware does not support
+ * the diag288 watchdog.
+ * User action:
+ * Confirm that the diag288 watchdog is supported in your environment.
+ * Use a watchdog that is supported in your environment.
+ */
+
+/*?
+ * Text: "The watchdog was activated\n"
+ * Severity: Informational
+ * Description:
+ * The diag288 watchdog is operational and counting.
+ * If the watchdog times out, the watchdog device driver will trigger the configured timeout action.
+ * User action:
+ * None.
+ */
+
+/*?
+ * Text: "Linux runs in an environment that does not support the diag288 watchdog\n"
+ * Severity: Error
+ * Description:
+ * The diag288 watchdog requires an LPAR or z/VM environment.
+ * You cannot use the diag288 watchdog.
+ * User action:
+ * Run your Linux instance in LPAR mode or as a z/VM guest
+ * or use a watchdog that is supported in your environment.
+ */
+
+/*?
+ * Text: "The watchdog cannot be deactivated\n"
+ * Severity: Error
+ * Description:
+ * Diagnose instruction 0x288 was called to deactivate the diag288 watchdog.
+ * The diagnose call returned an error that cannot be handled by the device driver.
+ * The watchdog stays active and a watchdog timeout will trigger the configured timeout action.
+ * The diag288 watchdog device driver might intentionally be configured to prevent deactivation.
+ * User action:
+ * You can configure the diag288 watchdog device driver such that it can be deactivated.
+ * If the diag288 device driver has been compiled as a separate module, diag288_wdt, reload the module
+ * without specifying the 'nowayout' module parameter.
+ * If the diag288 device driver has been compiled into your kernel,
+ * reboot Linux without specifying the 'diag288.nowayout' kernel parameter'.
+ */
+
+/*?
+ * Text: "The watchdog timer cannot be started or reset\n"
+ * Severity: Error
+ * Description:
+ * Diagnose instruction 0x288 was called to start the diag288 watchdog or to set timer back to zero.
+ * The diagnose call returned an error that cannot be handled by the device driver.
+ * The watchdog stays inactive or becomes inactive.
+ * User action:
+ * Contact your support organization.
+ */
+
+/*?
+ * Text: "Linux cannot be suspended while the watchdog is in use\n"
+ * Severity: Error
+ * Description:
+ * The watchdog must not time out while Linux is suspended.
+ * Therefore, the diag288 watchdog device driver prevents Linux from being suspended
+ * while the watchdog is in use.
+ * User action:
+ * i) Stop the watchdog application. ii) If the problem persists, close the watchdog
+ * device node by issuing 'echo V > /dev/watchdog'.
+ * iii) If the device driver still prevents Linux from being suspended,
+ * contact your support organization.
+ */
+
+/*?
+ * Text: "The watchdog device driver detected a z/VM environment\n"
+ * Severity: Informational
+ * Description:
+ * The diag288 watchdog device driver is running on a z/VM guest.
+ * If the watchdog is activated, it will run in z/VM mode.
+ * User action:
+ * None.
+ */
+
+/*?
+ * Text: "The watchdog device driver detected an LPAR environment\n"
+ * Severity: Informational
+ * Description:
+ * The diag288 watchdog device driver is running on a Linux instance in LPAR mode.
+ * If the watchdog is activated, it will run in LPAR mode.
+ * User action:
+ * None.
+ */
--- a/Documentation/kmsg/s390/vmwatchdog
+++ /dev/null
@@ -1,25 +0,0 @@
-/*?
- * Text: "The system cannot be suspended while the watchdog is in use\n"
- * Severity: Error
- * Description:
- * A program is currently using the vmwatchdog device node. The watchdog
- * device driver prevents the system from being suspended while the watchdog
- * device is in use.
- * User action:
- * If you want to suspend the system, find out which program uses the watchdog
- * device. Stop the program or reconfigure it to not use the watchdog.
- */
-
-
-/*?
- * Text: "The system cannot be suspended while the watchdog is running\n"
- * Severity: Error
- * Description:
- * The watchdog must not time out during hibernation. The watchdog
- * device driver prevents the system from being suspended while the watchdog
- * timer is running.
- * User action:
- * If you want to suspend the system, stop the watchdog, for example, by entering
- * the command: 'echo V > /dev/vmwatchdog'. Alternatively, stop the program that
- * uses the watchdog or reconfigure the program to not use the watchdog.
- */
